flang/lib/Parser/io-parsers.cpp

```diff
@@ -23,8 +23,12 @@ namespace Fortran::parser {
 // R905 char-variable -> variable
 // "char-variable" is attempted first since it's not type constrained but
 // syntactically ambiguous with "file-unit-number", which is constrained.
+// Note, "file-unit-number" is replaced by "expr" to allow for better
+// error messages.
 TYPE_PARSER(construct<IoUnit>(variable / lookAhead(space / ",);\n"_ch)) ||
-    construct<IoUnit>(fileUnitNumber) || construct<IoUnit>(star))
+    construct<IoUnit>(
```

Comment:
Would not the other uses of `fileUnitNumber` in this file also benefit from this change?  If so, consider modifying the parser for `FileUnitNumber` itself.

Reply:
When I was looking into it the other uses of fileUnitNumber in this file don't have the same overlap with a character variable that can lead to this issue, and thus I don't think they would benefit from this.